### PR TITLE
Add unit test cases for vfio-ioctls

### PR DESCRIPTION
**Pull request by jiangliu**
This PR prepares for publishing the vfio-ioctls by:
1) add a mock test framework
2) add unit test cases
3) support user mode device driver in addition to kvm/mshv, same goal as #33 
4) improve README.md

### CODE REVIEW

File: crates/vfio-ioctls/src/lib.rs [hunk 1]
```diff
@@ -57,12 +57,136 @@
...
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            VfioError::OpenContainer(e) => {
+                write!(f, "failed to open /dev/vfio/vfio container: {}", e)
+            }
+            VfioError::OpenGroup(e, ref p) => {
+                write!(f, "failed to open /dev/vfio/{} group: {}", p, e)
+            }
+            VfioError::GetGroupStatus => write!(f, "failed to get Group Status"),
+            VfioError::GroupViable => write!(f, "group is inviable"),
```
Comment (sboeuf) [inline crates/vfio-ioctls/src/lib.rs]:
```suggestion
            VfioError::GroupViable => write!(f, "group is not viable"),
```

File: crates/vfio-ioctls/src/lib.rs [hunk 2]
```diff
@@ -57,12 +57,136 @@
...
+            VfioError::VfioApiVersion => write!(
+                f,
+                "vfio API version doesn't match with VFIO_API_VERSION defined in vfio-bindings"
+            ),
+            VfioError::VfioExtension => write!(f, "failed to check VFIO extension"),
+            VfioError::VfioInvalidType => write!(f, "invalid VFIO type"),
+            VfioError::VfioType1V2 => {
+                write!(f, "container dones't support VfioType1V2 IOMMU driver type")
```
Comment (sboeuf) [inline crates/vfio-ioctls/src/lib.rs]:
```suggestion
                write!(f, "container doesn't support VfioType1V2 IOMMU driver type")
```

File: crates/vfio-ioctls/src/lib.rs [hunk 3]
```diff
@@ -70,110 +71,55 @@ pub use vfio_device::{
...
 pub enum VfioError {
-    OpenContainer(io::Error),
-    OpenGroup(io::Error, String),
+    #[error("failed to open /dev/vfio/vfio container: {0}")]
+    OpenContainer(#[source] io::Error),
+    #[error("failed to open /dev/vfio/{1} group: {0}")]
+    OpenGroup(#[source] io::Error, String),
+    #[error("failed to get Group Status")]
     GetGroupStatus,
+    #[error("group is inviable")]
```
Comment (sboeuf) [inline crates/vfio-ioctls/src/lib.rs:84]:
```suggestion
    #[error("group is not viable")]
```

File: crates/vfio-ioctls/src/lib.rs [hunk 4]
```diff
@@ -70,110 +71,55 @@ pub use vfio_device::{
...
     GroupViable,
+    #[error("vfio API version doesn't match with VFIO_API_VERSION defined in vfio-bindings")]
     VfioApiVersion,
+    #[error("failed to check VFIO extension")]
     VfioExtension,
+    #[error("invalid VFIO type")]
     VfioInvalidType,
+    #[error("container dones't support VfioType1V2 IOMMU driver type")]
```
Comment (sboeuf) [inline crates/vfio-ioctls/src/lib.rs:92]:
```suggestion
    #[error("container doesn't support VfioType1V2 IOMMU driver type")]
```

File: crates/vfio-ioctls/Cargo.toml [hunk 5]
```diff
@@ -17,6 +17,7 @@ byteorder = ">=1.2.1"
 log = "0.4"
 kvm-bindings = { version = "~0", optional = true }
 kvm-ioctls = { version = "~0", optional = true }
+thiserror = "1.0"
```
Comment (sboeuf) [inline crates/vfio-ioctls/Cargo.toml:23]: Can't we relax the requirement here with `thiserror = ">=1.0"`?

File: crates/vfio-ioctls/src/vfio_device.rs [hunk 6]
```diff
@@ -905,7 +905,8 @@ impl VfioDevice {
             .irqs
             .get(&irq_index)
             .ok_or(VfioError::VfioDeviceSetIrq)?;
-        if irq.count == 0 {
+        // Currently the VFIO driver only support MASK/UNMASK INTX, so count is hard-coded to 1.
+        if irq.count == 0 || irq.count != 1 {
```
Comment (sboeuf) [inline crates/vfio-ioctls/src/vfio_device.rs]: This should be in a separate commit.

File: crates/vfio-ioctls/src/vfio_device.rs [hunk 7]
```diff
@@ -791,7 +791,7 @@ impl VfioDevice {
             .irqs
             .get(&irq_index)
             .ok_or(VfioError::VfioDeviceSetIrq)?;
-        if irq.count < vector {
+        if irq.count <= vector {
             return Err(VfioError::VfioDeviceSetIrq);
```
Comment (sboeuf) [inline crates/vfio-ioctls/src/vfio_device.rs:807]: This is a general comment about `VfioError::VfioDeviceSetIrq` being used in too many places, making the error type useless because it's almost impossible to identify where the error came from.
Since you're improving the overall code with this PR, would you mind adding some new error types in order to make the error reporting better?

File: crates/vfio-ioctls/Cargo.toml [hunk 8]
```diff
@@ -14,6 +14,7 @@ mshv = ["mshv-ioctls", "mshv-bindings"]
 
 [dependencies]
 byteorder = ">=1.2.1"
+libc = "0.2.108"
```
Comment (sboeuf) [inline crates/vfio-ioctls/Cargo.toml:19]: Can't we relax this dependency?

File: crates/vfio-ioctls/src/vfio_device.rs [hunk 9]
```diff
@@ -906,7 +826,7 @@ impl VfioDevice {
             .get(&irq_index)
             .ok_or(VfioError::VfioDeviceSetIrq)?;
         // Currently the VFIO driver only support MASK/UNMASK INTX, so count is hard-coded to 1.
-        if irq.count == 0 || irq.count != 1 {
+        if irq.count == 0 || irq.count != 1 || irq.index != 0 {
```
Comment (sboeuf) [inline crates/vfio-ioctls/src/vfio_device.rs:902]: Can you please fix this in a separate commit all at once.

File: crates/vfio-ioctls/src/vfio_device.rs [hunk 10]
```diff
@@ -388,12 +390,21 @@ impl VfioGroup {
...
+                .open(tmp_file.as_path())
+                .map_err(|e| VfioError::OpenGroup(e, id.to_string()))?
+        } else {
+            let group_path = Path::new("/dev/vfio").join(id.to_string());
+            OpenOptions::new()
+                .read(true)
+                .write(true)
+                .open(&group_path)
+                .map_err(|e| VfioError::OpenGroup(e, id.to_string()))?
+        };
```
Comment (sboeuf) [inline crates/vfio-ioctls/src/vfio_device.rs:407]: I'm not a fan of introducing test specific code in the actual codepath that's being used for non-test cases. I'd rather have two functions `new()`, one for each case, even if this duplicates a little bit more code.

File: README.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# Safe wrappers for VFIO
+
+## Design
+
+This repository provides safe wrappers over the
+[VFIO driver framework](https://www.kernel.org/doc/Documentation/vfio.txt).
+
+Many modern system now provide DMA and interrupt remapping facilities to help ensure I/O devices behave within the boundaries they’ve been allotted. This includes x86 hardware with AMD-Vi and Intel VT-d, POWER systems with Partitionable Endpoints (PEs) and embedded PowerPC systems such as Freescale PAMU. The VFIO driver is an IOMMU/device agnostic framework for exposing direct device access to userspace, in a secure, IOMMU protected environment. In other words, the VFIO framework allows safe, non-privileged, userspace drivers.
```
Comment (sboeuf) [inline README.md:8]:
```suggestion
Many modern systems now provide DMA and interrupt remapping facilities to help ensure I/O devices behave within the boundaries they’ve been allotted. This includes x86 hardware with AMD-Vi and Intel VT-d, POWER systems with Partitionable Endpoints (PEs) and embedded PowerPC systems such as Freescale PAMU. The VFIO driver is an IOMMU/device agnostic framework for exposing direct device access to userspace, in a secure, IOMMU protected environment. In other words, the VFIO framework allows safe, non-privileged, userspace drivers.
```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+Why do we want that? Virtual machines often make use of direct device access (“device assignment”) when configured for the highest possible I/O performance. From a device and host perspective, this simply turns the VM into a userspace driver, with the benefits of significantly reduced latency, higher bandwidth, and direct use of bare-metal device drivers.
+
+Devices are the main target of any I/O driver. Devices typically create a programming interface made up of I/O access, interrupts, and DMA. Without going into the details of each of these, DMA is by far the most critical aspect for maintaining a secure environment as allowing a device read-write access to system memory imposes the greatest risk to the overall system integrity.
```
Comment (sboeuf) [inline README.md:12]:
```suggestion
Devices are the main target of any I/O driver. Devices typically create a programming interface made up of I/O accesses, interrupts, and DMA. Without going into the details of each of these, DMA is by far the most critical aspect for maintaining a secure environment as allowing a device read-write access to system memory imposes the greatest risk to the overall system integrity.
```

File: README.md [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+To help mitigate this risk, many modern IOMMUs now incorporate isolation properties into what was, in many cases, an interface only meant for translation (ie. solving the addressing problems of devices with limited address spaces). With this, devices can now be isolated from each other and from arbitrary memory access, thus allowing things like secure direct assignment of devices into virtual machines.
+
+While for the most part an IOMMU may have device level granularity, any system is susceptible to reduced granularity. The IOMMU API therefore supports a notion of IOMMU groups. A group is a set of devices which is isolatable from all other devices in the system. Groups are therefore the unit of ownership used by VFIO.
```
Comment (sboeuf) [inline README.md:16]:
```suggestion
While for the most part an IOMMU may have device level granularity, any system is susceptible to expose a reduced granularity. The IOMMU API therefore supports a notion of IOMMU groups. A group is a set of devices which is isolated from all other devices in the system. Groups are therefore the unit of ownership used by VFIO.
```

File: README.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+While the group is the minimum granularity that must be used to ensure secure user access, it’s not necessarily the preferred granularity. In IOMMUs which make use of page tables, it may be possible to share a set of page tables between different groups, reducing the overhead both to the platform (reduced TLB thrashing, reduced duplicate page tables), and to the user (programming only a single set of translations). For this reason, VFIO makes use of a container class, which may hold one or more groups. A container is created by simply opening the /dev/vfio/vfio character device.
+
+## Usage
+This repository provides two crates to use the VFIO framework, please refer to crate documentations for detail information.
+- [vfio-bindings](https://github.com/jiangliu/vfio-ioctls/tree/ioctls/crates/vfio-bindings): a rust FFI bindings to VFIO generated using [bindgen](https://crates.io/crates/bindgen).
```
Comment (sboeuf) [inline README.md:22]:
```suggestion
- [vfio-bindings](https://github.com/rust-vmm/vfio-ioctls/tree/ioctls/crates/vfio-bindings): a rust FFI bindings to VFIO generated using [bindgen](https://crates.io/crates/bindgen).
```

File: README.md [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+- [vfio-ioctls](https://github.com/jiangliu/vfio-ioctls/tree/ioctls/crates/vfio-ioctls): a group of safe wrappers over the [VFIO APIs](https://github.com/torvalds/linux/blob/master/include/uapi/linux/vfio.h).
```
Comment (sboeuf) [inline README.md:23]:
```suggestion
- [vfio-ioctls](https://github.com/rust-vmm/vfio-ioctls/tree/ioctls/crates/vfio-ioctls): a group of safe wrappers over the [VFIO APIs](https://github.com/torvalds/linux/blob/master/include/uapi/linux/vfio.h).
```

File: crates/vfio-ioctls/README.md [hunk 16]
```diff
@@ -1,41 +1,68 @@
...
+This framework is used for multiple devices, such as GPUs, network adapters, and compute
+accelerators. With direct device access, virtual machines or user space applications have
+direct access to the physical device.
 
-Some questions that might help in writing this section:
-- What is the purpose of this crate?
-- What are the main components of the crate? How do they interact which each
-  other?
+The VFIO framework is originally developed on Linux system, and later Microsoft HyperV
+technology provides an compatible implementation. So the VFIO framework is supported
```
Comment (sboeuf) [inline crates/vfio-ioctls/README.md:13]:
```suggestion
technology provides a compatible implementation. Therefore the VFIO framework is supported
```

**Review by sboeuf — COMMENTED** [review]
Very nice PR :+1: 
I've tested it with Cloud Hypervisor and I confirm it works correctly :)

File: crates/vfio-ioctls/Cargo.toml (hunk 8 above)
Comment (jiangliu) [reply crates/vfio-ioctls/Cargo.toml:19]: libc = ">=0.2.39" should work most of our cases.

**Review by sboeuf — COMMENTED** [review]
@jiangliu 
Thanks for the quick update of your PR.
I don't think you addressed the comments on `libc` and `thiserror` dependencies. Just wanted to have your thoughts on these.

**Comment by jiangliu** [comment]
> @jiangliu Thanks for the quick update of your PR. I don't think you addressed the comments on `libc` and `thiserror` dependencies. Just wanted to have your thoughts on these.

It has been updated by the last patch, I will move the modification to the right patch.

**Review by sboeuf — DISMISSED** [review]

File: crates/vfio-ioctls/src/fam.rs [hunk 17]
```diff
@@ -33,7 +40,50 @@ fn vec_with_size_in_bytes<T: Default>(size_in_bytes: usize) -> Vec<T> {
 /// as a `Foo`. The remaining memory in the `Vec<Foo>` is for `entries`, which must be contiguous
 /// with `Foo`. This function is used to make the `Vec<Foo>` with enough space for `count` entries.
 pub(crate) fn vec_with_array_field<T: Default, F>(count: usize) -> Vec<T> {
-    let element_space = count * size_of::<F>();
-    let vec_size_bytes = size_of::<T>() + element_space;
+    let element_space = match count.checked_mul(size_of::<F>()) {
```
Comment (andreeaflorescu) [inline crates/vfio-ioctls/src/fam.rs:43]: Why is this still needed? I noticed you're using the fam struct from vmm-sys-util as well.

Comment (jiangliu) [reply crates/vfio-ioctls/src/fam.rs:43]: It may be replaced by FamStruct, I hope do that with a separate PR:)

Comment (jiangliu) [reply crates/vfio-ioctls/src/fam.rs:43]: I have double checked the code, and it seems FamStruct doesn't help much here. The structure used with `vec_with_array_field` have conflicts with FamStruct's assumption.
1) those structs has field "argsz" (in bytes) instead of `n_elements`
2) the following elements may be different type/size.

Need more thinking about whether it worth to switching to FamStruct.

File: crates/vfio-ioctls/README.md [hunk 18]
```diff
@@ -1,41 +1,68 @@
...
+exposing direct device access to user space in a secure, IOMMU-protected environment.
+This framework is used for multiple devices, such as GPUs, network adapters, and compute
+accelerators. With direct device access, virtual machines or user space applications have
+direct access to the physical device.
 
-Some questions that might help in writing this section:
-- What is the purpose of this crate?
-- What are the main components of the crate? How do they interact which each
-  other?
+The VFIO framework is originally developed on Linux system, and later Microsoft HyperV
```
Comment (liuw) [inline crates/vfio-ioctls/README.md:12]: `Microsoft HyperV` should be `Microsoft Hypervisor`

File: crates/vfio-ioctls/README.md [hunk 19]
```diff
@@ -1,41 +1,68 @@
...
+technology provides a compatible implementation. Therefore the VFIO framework is supported
+by both Linux and Windows HyperV.
```
Comment (liuw) [inline crates/vfio-ioctls/README.md:14]: Same here.

File: crates/vfio-ioctls/README.md [hunk 20]
```diff
@@ -1,41 +1,68 @@
...
+- `VfioDevice`: a wrapper over a VFIO device object, provide methods to access the underlying
+  hardware device.
+- `VfioIrq/VfioRegion`: describes capabilities/resources about a `VfioDevice` object. 
 
 ## Usage
 
-TODO: This section describes how the crate is used.
+The `vfio-ioctls` crate may be used to support following usage scenarios:
+- Direct device assignment to virtual machine based on Linux KVM, with default features.
+- Direct device assignment to virtual machine based on Microsoft HyperV, with `--no-default-features --features=mshv`.
```
Comment (liuw) [inline crates/vfio-ioctls/README.md:27]: Same here.

File: crates/vfio-ioctls/src/vfio_device.rs [hunk 21]
```diff
@@ -309,13 +254,124 @@ impl VfioContainer {
...
+
+            self.device_fd
+                .set_device_attr(&dev_attr)
+                .map_err(VfioError::SetDeviceAttr)
+        }
+    }
+}
+
+#[cfg(all(feature = "mshv", not(feature = "kvm"), not(test)))]
+// Methods to support the Microsoft HyperV hypervisor.
```
Comment (liuw) [inline crates/vfio-ioctls/src/vfio_device.rs:304]: Just use `Microsoft Hypervisor` here.

**Review by liuw — CHANGES_REQUESTED** [review]

**Review by sboeuf — APPROVED** [review]

**Review by sameo — APPROVED** [review]

**Comment by sboeuf** [comment]
@jiangliu now this is merged, are you planning on updating the repository name to `vfio`?

**Comment by liuw** [comment]
> @jiangliu now this is merged, are you planning on updating the repository name to `vfio`?

Can we make a new copy? Making the old one disappear will break builds and make code archaeology much harder.

**Comment by jiangliu** [comment]
> @jiangliu now this is merged, are you planning on updating the repository name to `vfio`?

@andreeaflorescu any suggestions here? How about creating a new vfio repo and archives the vfio-bindings/vfio-ioctls repos? We could clone from vfio-ioctls into vfio repo to keep all history info.

**Comment by andreeaflorescu** [comment]
I created https://github.com/rust-vmm/vfio. You can set it up as discussed here.
